qf_lib/data_providers/sp_global/sp_data_provider.py

```diff
@@ -0,0 +1,359 @@
+                                                                                func.current_date())),
+                                      divexrate.pricedate == self.ciqpriceequity.pricingdate)) \
+                .filter(self.ciqtradingitem.tradingitemid.in_(tickers)) \
+                .filter(peexrate.snapid == self.exchange_rate_snap_id) \
+                .filter(divexrate.snapid == self.exchange_rate_snap_id) \
+                .filter(self.ciqpriceequity.pricingdate.between(start_datetime.date(), end_datetime.date())) \
+                .order_by(self.ciqpriceequity.tradingitemid, self.ciqpriceequity.pricingdate,
+                          self.ciqiadividendchain.startdate).statement
+
+            timestamp_string = self.ciqpriceequity.pricingdate.key
+            data_frame = QFDataFrame(pd.read_sql(query, con=session.bind))
+            data_frame = data_frame.replace([None], np.nan)
+            data_frame = data_frame.drop_duplicates(subset=["tradingitemid", timestamp_string], keep="last")
+            data_frame[timestamp_string] = pd.to_datetime(data_frame[timestamp_string])
+            data_frame = data_frame.set_index(["tradingitemid", timestamp_string])
+        return data_frame
+
+    def _fields_to_adjust(self) -> Set[SPField]:
+        """
+        Return a list of fields available in the Data Provider which are either adjustable or can be translated into
+        a different currency (USD).
+        """
+        return {SPField.AskPrice, SPField.BidPrice, SPField.LowPrice, SPField.HighPrice, SPField.OpenPrice,
+                SPField.ClosePrice}
+
+    def _get_futures_chain_dict(self, tickers: Union[FutureTicker, Sequence[FutureTicker]],
+                                expiration_date_fields: Union[str, Sequence[str]]) -> Dict[FutureTicker, QFDataFrame]:
+        raise NotImplementedError()
+
+    def expiration_date_field_str_map(self, ticker: Ticker = None) -> Dict[ExpirationDateField, str]:
+        raise NotImplementedError()
+
+    def _get_div_adjusted_prices(self, data_frame, tids_and_dates, div_adj_factor):
+        indices_union = data_frame.index.union(div_adj_factor.index)
+        data_frame = data_frame.reindex(index=indices_union)
+        div_adj_factor = div_adj_factor.reindex(index=indices_union).groupby(level=0).ffill()
+
+        tids_and_dates = div_adj_factor.index.intersection(tids_and_dates)
+        vals = div_adj_factor.groupby(level=0).shift(1).loc[tids_and_dates] / div_adj_factor.loc[tids_and_dates]
+        ratio = vals.groupby(level=0).cumprod().reindex_like(div_adj_factor, method="bfill") \
+            .groupby(level=0).shift(-1).fillna(1.0)
+
+        fields_to_adjust_str = {sp_field.value for sp_field in self._fields_to_adjust()}
+        fields = list(fields_to_adjust_str.intersection(data_frame.columns))
+        data_frame.loc[:, fields] = data_frame.loc[:, fields].mul(ratio['divadjfactor'], axis=0)
+        return data_frame.dropna(how="all")
```

Comment:
This is a mess, let's try to rewrite a bit nicer.
Or maybe we should create a separate class for an SP data provider with the div adjusted prices?
This way we can have a minimal class to just get the prices, and then add more functionality on top of it in another class.